### PR TITLE
Implement the lineage event logger

### PR DESCRIPTION
**Pull request by fm100**
Signed-off-by: Minkyu Park <minkyu.park.200@gmail.com>

### Problem
As proposed in the design of the client proxy, this PR implements the event logger that stores validated lineage events to the local database implemented with sqlite3.

### Solution
For the new lineage event, it simply adds the event to the lineage event partition table. When the partition hits its limit, it rotates the partition by creating a new partition table. The partition limit is configured in the configuration file `proxy.env` and overridible by setting environment variables.

> **Note:** All schema changes require discussion. Please [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) for context.

- [ ] Your change modifies the [core](https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.json) OpenLineage model
- [ ] Your change modifies one or more OpenLineage [facets](https://github.com/OpenLineage/OpenLineage/tree/main/spec/facets)

If you're contributing a new integration, please specify the scope of the integration and how/where it has been tested (e.g., Apache Spark integration supports `S3` and `GCS` filesystem operations, tested with AWS EMR).

### Checklist

- [ ] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [ ] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [ ] Your changes are accompanied by tests (_if relevant_)
- [ ] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [ ] You've updated the [`CHANGELOG.md`](https://github.com/OpenLineage/OpenLineage/blob/main/CHANGELOG.md) with details about your change under the "Unreleased" section (_if relevant, depending on the change, this may not be necessary_)
- [ ] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)
- [ ] You've added a [header](https://github.com/OpenLineage/OpenLineage/tree/main/.github/header_templates.md) to source files (_if relevant_)

----
SPDX-License-Identifier: Apache-2.0\
Copyright 2018-2023 contributors to the OpenLineage project

### CODE REVIEW

File: proxy/client/logger/logger.go [hunk 1]
```diff
@@ -4,33 +4,50 @@
...
+func (l *EventLogger) Log(lineageEvent string) error {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	p, err := l.storage.GetCurrentPartition()
+	if err != nil {
+		return err
+	}
+
+	continuePartition := p.Size+uint64(len(lineageEvent)) < l.partitionLimitBytes &&
```
Comment (wslulciuc) [inline proxy/client/logger/logger.go:34]: _Minor_: A comment on the calculation logic would help with readability.

File: proxy/client/services/lineage/service.go [hunk 2]
```diff
@@ -21,17 +22,20 @@ type ILineageService interface {
...
-	if err != nil {
+	if err := s.validator.Validate(lineageEvent); err != nil {
 		go s.failedEventHandler.Handle(lineageEvent)
 		return http.StatusBadRequest
 	}
 
-	s.eventLogger.Log(lineageEvent)
+	if err := s.eventLogger.Log(lineageEvent); err != nil {
+		go s.failedEventHandler.Handle(lineageEvent)
+		return http.StatusInternalServerError
```
Comment (wslulciuc) [inline proxy/client/services/lineage/service.go:32]: Do we want to return a `500` to the caller? Or, can we return a more helpful message?

File: proxy/client/config/config.go [hunk 3]
```diff
@@ -1,22 +1,23 @@
-
 // Copyright 2018-2023 contributors to the OpenLineage project
 // SPDX-License-Identifier: Apache-2.0
 
-
 package config
 
 import "github.com/spf13/viper"
 
 type Config struct {
```
Comment (wslulciuc) [inline proxy/client/config/config.go:8]: Can `Config` define reasonable defaults?

File: proxy/client/storage/storage.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+	_ "github.com/mattn/go-sqlite3"
+	"math/rand"
+	"time"
+)
+
+type Storage struct {
+	db *sql.DB
+}
+
+type Partition struct {
```
Comment (wslulciuc) [inline proxy/client/storage/storage.go:19]: Migration [`1_initial_schema.up.sql`](https://github.com/OpenLineage/OpenLineage/blob/f314393b708a8544fea4d9dcb18f10ae233d1f28/proxy/client/storage/migrations/1_initial_schema.up.sql) doesn't have `RowID` defined?

File: proxy/client/storage/storage.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+	Size      uint64
+	CreatedAt time.Time
+	IsCurrent bool
+}
+
+// IStorage is an interface for interaction with local storage database
+type IStorage interface {
+	GetCurrentPartition() (*Partition, error)
+	RotatePartition() (*Partition, error)
+	InsertLineageEvent(*Partition, string) error
```
Comment (wslulciuc) [inline proxy/client/storage/storage.go:31]: _Minor_: I would be consistent with method names (_wordy_ vs _less_). That is, in `EventLogger` you define:

```
EventLogger.Log()
```

And for the storage you have:

```
IStorage.InsertLineageEvent()
```

I think a method naming strategy would help keep things consistent. I would either be _wordy_ with `EventLogger.LogLineageEvent()` or _less_ and go with `IStorage.Insert()`

File: proxy/client/storage/storage.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	_, err = tx.Exec(`
+			INSERT INTO partitions(name, size, created_at, is_current)
+			VALUES(?, ?, ?, ?)
+		`, tableName, 0, time.Now(), true)
+	return nil
+}
+
+func generatePartitionSuffix() string {
```
Comment (wslulciuc) [inline proxy/client/storage/storage.go:149]: Mind if we document the logic of `generatePartitionSuffix()` with some example partitions?

Comment (wslulciuc) [reply proxy/client/storage/storage.go:149]: _Minor_: Have we considered partitioning by time or similar?

```
lineage_events_partition_20230201
```

File: proxy/client/storage/storage.go [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+	row := db.QueryRow("SELECT rowid, name, size, created_at, is_current FROM partitions WHERE is_current=true")
+	if err := row.Scan(&rowID, &name, &size, &createdAt, &isCurrent); err != nil {
+		return nil, err
+	}
+
+	return &Partition{RowID: rowID, Name: name, Size: size, CreatedAt: createdAt, IsCurrent: isCurrent}, nil
+}
+
+func createPartitionTable(tx *sql.Tx) error {
+	tableName := fmt.Sprintf("lineage_events_partition_%s", generatePartitionSuffix())
```
Comment (wslulciuc) [inline proxy/client/storage/storage.go:131]: Outside of this method, it's not clear how `lineage_events_partition_*` is used. Can we adds docs to the method?

**Comment by harels** [comment]
closing this PR since we have the fluentd proxy. feel free to reopen if anyone would like to pick this up.

**Pull request closed without merging**